examples/src/margotree.c

```diff
@@ -0,0 +1,878 @@
+    MPI_Barrier(MPI_COMM_WORLD);
+
+    return ret;
+}
+
+int main(int argc, char* argv[])
+{
+  MPI_Init(&argc, &argv);
+
+  MPI_Comm_rank(MPI_COMM_WORLD, &glb_rank);
+  MPI_Comm_size(MPI_COMM_WORLD, &glb_ranks);
+
+  glb_addrs   = (char*) malloc(glb_ranks * 128);
+  glb_servers = (server_info_t*) malloc(glb_ranks * sizeof(server_info_t));
+
+  glb_mid = setup_remote_target();
+
+  glb_filesize_request_id =
+        MARGO_REGISTER(glb_mid, "filesize_request_rpc",
+                       filesize_request_in_t, filesize_request_out_t,
+                       filesize_request_rpc);
+
+  glb_filesize_response_id =
+        MARGO_REGISTER(glb_mid, "filesize_response_rpc",
+                       filesize_response_in_t, filesize_response_out_t,
+                       filesize_response_rpc);
+
+  glb_filesize_request_noreply_id =
+        MARGO_REGISTER(glb_mid, "filesize_request_noreply_rpc",
+                       filesize_request_in_t, void,
+                       filesize_request_noreply_rpc);
+  margo_registered_disable_response(glb_mid, glb_filesize_request_noreply_id, HG_TRUE);
+
+  glb_filesize_response_noreply_id =
+        MARGO_REGISTER(glb_mid, "filesize_response_noreply_rpc",
+                       filesize_response_in_t, void,
+                       filesize_response_noreply_rpc);
+  margo_registered_disable_response(glb_mid, glb_filesize_response_noreply_id, HG_TRUE);
+
+  margo_connect_servers(glb_mid);
+
+  for (int reply = 0; reply < 2; reply++) {
+  for (int degree = 2; degree < 9; degree++) {
+
+  MPI_Barrier(MPI_COMM_WORLD);
+
+  if (glb_rank == 0) {
+    double start = MPI_Wtime();
+
+    int root = 0;
+    int tag = glb_rank;
+    glb_st = state_filesize_alloc(root, degree, reply, -1, tag);
+
+    /* lock structure until we can wait on it */
+    if (glb_ranks > 1) {
+        /* init mutex and lock before we use them,
+         * these are only needed on the root server */
+        ABT_mutex_create(&glb_st->mutex);
+        ABT_cond_create(&glb_st->cond);
+
+        /* lock the mutex */
+        ABT_mutex_lock(glb_st->mutex);
+    }
+
+    /* start the reduction to get the file size */
+    filesize_request_forward(glb_st);
+
+    /* wait on signal that reduction has completed */
+    if (glb_ranks > 1) {
+        ABT_cond_wait(glb_st->cond, glb_st->mutex);
+    }
+
+    /* have result at this point, get it */
+    size_t filesize = glb_st->filesize;
+    //printf("BUCKEYES got a filesize of %llu\n",
+    //    (unsigned long long) glb_st->filesize);
+    //fflush(stdout);
+
+    /* release lock and free state */
+    if (glb_ranks > 1) {
+        ABT_mutex_unlock(glb_st->mutex);
+    }
+
+    state_filesize_free(&glb_st);
+
+    double end = MPI_Wtime();
+    printf("procs=%d degree=%d reply=%d secs=%f\n", glb_ranks, degree, reply, (end - start));
+  }
+
+  MPI_Barrier(MPI_COMM_WORLD);
+
+  double start = MPI_Wtime();
+
+  int times = 100;
+  for (int i = 0; i < times; i++) {
```

Comment:
@sandrain , after that I then executed 100 bcasts in a row here and reported the average cost on an iteration.  In this case, all of them were started from root 0, and the next bcast didn't start until the previous one completed.